src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,361 @@
+        let operator_waker = TimelyWaker {
+            activator,
+            active: AtomicBool::new(false),
+            task_ready: AtomicBool::new(false),
+        };
+
+        OperatorBuilder {
+            builder,
+            shared_frontiers: Default::default(),
+            registered_wakers: Default::default(),
+            operator_waker: Arc::new(operator_waker),
+            input_handles: Default::default(),
+        }
+    }
+
+    /// Adds a new input, returning the async input handle to use.
+    pub fn new_input<D: Container, P>(
+        &mut self,
+        stream: &StreamCore<G, D>,
+        pact: P,
+    ) -> AsyncInputHandle<G::Timestamp, D, P::Puller>
+    where
+        P: ParallelizationContractCore<G::Timestamp, D>,
+    {
+        let connection =
+            vec![Antichain::from_elem(Default::default()); self.builder.shape().outputs()];
+        self.new_input_connection(stream, pact, connection)
+    }
+
+    /// Adds a new input with connection information, returning the async input handle to use.
+    ///
+    /// The `connection` parameter contains promises made by the operator for each of the existing
+    /// *outputs*, that any timestamp appearing at the input, any output timestamp will be greater
+    /// than or equal to the input timestamp subjected to a `Summary` greater or equal to some
+    /// element of the corresponding antichain in `connection`.
+    ///
+    /// Commonly the connections are either the unit summary, indicating the same timestamp might
+    /// be produced as output, or an empty antichain indicating that there is no connection from
+    /// the input to the output.
+    pub fn new_input_connection<D: Container, P>(
+        &mut self,
+        stream: &StreamCore<G, D>,
+        pact: P,
+        connection: Vec<Antichain<<G::Timestamp as Timestamp>::Summary>>,
+    ) -> AsyncInputHandle<G::Timestamp, D, P::Puller>
+    where
+        P: ParallelizationContractCore<G::Timestamp, D>,
+    {
+        let index = self.builder.shape().inputs();
+        self.shared_frontiers
+            .borrow_mut()
+            .push((Antichain::from_elem(G::Timestamp::minimum()), false));
+
+        self.input_handles.set(self.input_handles.get() + 1);
+
+        AsyncInputHandle {
+            sync_handle: self.builder.new_input_connection(stream, pact, connection),
+            shared_frontiers: Rc::clone(&self.shared_frontiers),
+            reactor_registry: Rc::clone(&self.registered_wakers),
+            input_handles: Rc::clone(&self.input_handles),
+            index,
+        }
+    }
+
+    /// Adds a new output, returning the output handle and stream.
+    pub fn new_output<D: Container>(
+        &mut self,
+    ) -> (
+        OutputWrapper<G::Timestamp, D, TeeCore<G::Timestamp, D>>,
+        StreamCore<G, D>,
+    ) {
+        let connection =
+            vec![Antichain::from_elem(Default::default()); self.builder.shape().inputs()];
+        self.new_output_connection(connection)
```

Comment:
just call `self.builder.new_output()`, to make it more clear that for output we just forward on